File: solutions/java/binary-search/1/src/main/java/BinarySearch.java
import java.util.List;

class BinarySearch {
    List<Integer> items;
    
    BinarySearch(List<Integer> items) {
        this.items = items;
    }

    int indexOf(int item) throws ValueNotFoundException {
        if (items.size() == 0) {
            throw new ValueNotFoundException("Value not in array");
        }
        int mid = items.size() / 2;
        Integer midElem = items.get(mid);
        if (item < midElem) {
            BinarySearch rec = new BinarySearch(items.subList(0, mid));
            return rec.indexOf(item);
        } else if (item > midElem) {
            BinarySearch rec = new BinarySearch(items.subList(mid + 1, items.size()));
            return rec.indexOf(item) + mid + 1;
        } else {
            return mid;
        }
    }
}

File: solutions/kotlin/binary-search/1/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(list: List<Int>, item: Int): Int {
        val mid = list.size / 2
        val midElem = list.getOrNull(mid) ?: throw NoSuchElementException()
        return when {
            item < midElem -> search(list.take(mid), item)
            item > midElem -> mid + 1 + search(list.subList(mid+1, list.size), item)
            else -> mid
        }
    }
}
